### PR TITLE
Bump xunit and related fixes

### PR DESCRIPTION
**Pull request by sungam3r**
I suggest to say goodbye for netcoreapp3.1 and net5.0 since the-latest-and-greatest xunit does not support these TFMs, see #344 and #345.

изображение

### CODE REVIEW

File: src/GraphQLParser.Tests/Visitors/ASTVisitorTests.cs [hunk 1]
```diff
@@ -23,25 +23,25 @@ public void ASTVisitor_Should_Handle_Null()
...
 
     [Fact]
-    public void ASTVisitor_Should_Throw_On_Unknown_Node()
+    public async Task ASTVisitor_Should_Throw_On_Unknown_Node()
     {
         var visitor = new ASTVisitor<Context>();
         var context = new Context();
 
-        var ex = Should.Throw<NotSupportedException>(() => visitor.VisitAsync(new MySuperNode(), context).GetAwaiter().GetResult());
+        var ex = await Should.ThrowAsync<NotSupportedException>(async () => await visitor.VisitAsync(new MySuperNode(), context));
```
Comment (sungam3r) [inline src/GraphQLParser.Tests/Visitors/ASTVisitorTests.cs:31]: https://xunit.net/xunit.analyzers/rules/xUnit1031

File: src/GraphQLParser.Tests/Visitors/CountVisitorTests.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public async Task CountVisitor_Should_Count_Nodes(string text, int expectedCount
 
         var document = text.Parse();
 
-        await visitor.VisitAsync(document, context).ConfigureAwait(false);
+        await visitor.VisitAsync(document, context);
```
Comment (sungam3r) [inline src/GraphQLParser.Tests/Visitors/CountVisitorTests.cs:41]: https://xunit.net/xunit.analyzers/rules/xUnit1030

**Comment by Shane32** [comment]
> I suggest to say goodbye for netcoreapp3.1 and net5.0 since the-latest-and-greatest xunit does not support these TFMs, see #344 and #345.
> 
> изображение

I still use older TFMs in various applications and it does not hurt us to test for them. If you want to bump xunit, just make the dependency conditional on the tfm and use the older version for 3.1/5.0.

File: src/GraphQLParser.Tests/Visitors/CountVisitorTests.cs (hunk 2 above)
Comment (Shane32) [reply src/GraphQLParser.Tests/Visitors/CountVisitorTests.cs:41]: Please note that we had to add ConfigureAwait calls to DataLoader tests to make them work properly. If we make similar changes in the main project we will need to be sure multithreaded tests work reliably.

Comment (sungam3r) [reply src/GraphQLParser.Tests/Visitors/CountVisitorTests.cs:41]: I understand. It's a matter of balance. For these repo I think `.ConfigureAwait(false)` calls in tests are just noise.

**Comment by Shane32** [comment]
Note that if we do drop .NET 3.1 testing, we implicitly drop .NET Standard 2.1 testing. 👎

**Comment by Shane32** [comment]
And it seems that we don't test against the .NET Framework currently (a supported TFM), which means that we are not testing the .NET Standard 2.0 build. 👎

**Comment by sungam3r** [comment]
True. Is it worth to?

**Comment by Shane32** [comment]
If we have have different builds for .NET Standard 2.0/2.1 we should test against it.  It's not like it costs any dev time and CI testing is free.  It would be nice if we could test each of the 3 builds against .NET 6, but that would likely be quite a bit more complicated than simply testing .NET 4.8, .NET 3.1 and .NET 6.0.

I certainly have GraphQL applications that run on .NET Framework under a fully supported scenario, as I expect others do...

**Comment by sungam3r** [comment]
OK, I'll deal with this tomorrow making dependency conditional.

**Comment by Shane32** [comment]
> Arguably we could remove the .NET Standard 2.1 TFM from the codebase, using .NET Standard 2.0 as a fallback for older TFMs. I'm not prepared to do so at this time.

I'll also review this.  Perhaps there is little practical benefit to having a .NET Standard 2.1 TFM.

**Comment by Shane32** [comment]
Looking at the codebase, there isn't any code that wouldn't be exercised if we tested on both .NET Framework 4.8 and .NET 6, omitting .NET 3.1.  Again, I'd probably just run it on 3.1 anyway, but if we do eliminate it, at least all the code paths get tested.

I wouldn't eliminate the .NET Standard 2.1 TFM.  For one thing, it eliminates some unnecessary dependencies.

**Comment by Shane32** [comment]
We don't need to test against .NET 5.  It isn't a LTM release, and .NET 3.1 testing will cover the .NET Standard 2.1 binary.

File: src/GraphQLParser.Tests/Visitors/SDLPrinterFromParsedTextTests.cs [hunk 3]
```diff
@@ -924,11 +924,11 @@ public async Task SDLPrinter_Should_Print_BlockStrings(int number, string input,
         var document = (input + " scalar a").Parse();
 
         var printer = new SDLPrinter();
-        await printer.PrintAsync(document, writer).ConfigureAwait(false);
+        await printer.PrintAsync(document, writer);
         var renderedOriginal = writer.ToString();
 
-        var lines = renderedOriginal.Split(Environment.NewLine);
-        var renderedDescription = string.Join(Environment.NewLine, lines.SkipLast(1));
+        var lines = renderedOriginal.Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
```
Comment (sungam3r) [inline src/GraphQLParser.Tests/Visitors/SDLPrinterFromParsedTextTests.cs:930]: I had to change signatures because of net462 TFM that lacks of previous signatures.

File: src/GraphQLParser/Visitors/SDLPrinterExtensions.cs [hunk 4]
```diff
@@ -24,13 +24,28 @@ public static string Print(this SDLPrinter printer, ASTNode node)
...
+
     /// <summary>
     /// Prints the specified AST into the specified <see cref="MemoryStream"/> as a SDL document.
     /// If no encoding is specified, the document is written in UTF-8 format without a byte order mark.
     /// </summary>
     public static void Print(this SDLPrinter printer, ASTNode node, MemoryStream memoryStream, Encoding? encoding = null)
     {
-        using var streamWriter = new StreamWriter(memoryStream, encoding, -1 /* default */, true);
+        int bufferSize = -1;
+#if NET462
```
Comment (sungam3r) [inline src/GraphQLParser/Visitors/SDLPrinterExtensions.cs:38]: CoreCLR https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/IO/StreamWriter.cs#L94

NET Framework https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/io/streamwriter.cs#L140

Actually this method **does not work at all** in NET Framework for now throwing ArgumentOutOfRangeException. Caller can pass not-null encoding, but buffer size is always negative. 😮

Comment (sungam3r) [reply src/GraphQLParser/Visitors/SDLPrinterExtensions.cs:38]: I did not touch signatures to not make breaking changes.

Comment (sungam3r) [reply src/GraphQLParser/Visitors/SDLPrinterExtensions.cs:38]: As option - add new method with bufferSize or change existing arguments in v10 to not bother with additional net462 TFM and all jumping around.

Comment (Shane32) [reply src/GraphQLParser/Visitors/SDLPrinterExtensions.cs:38]: Looking at the MS docs, we should not be passing `null` to `encoding`, nor should we be passing `-1` to `bufferSize`.  What we should learn here is that if we do not follow specs, then the code can fail.  For example, what would occur if someone tried to run this on Mono or Xamrin or some other .NET Standard implementation?  We cannot assume that `-1` and `null` would work there any more than it does on .NET Framework.  I would remove the ifdef and leave the code, so that no matter if it is .NET Standard or .NET 6/7/8/9/etc, it will adhere to the current specification, which lists these exceptions:

> ArgumentNullException : path or encoding is null.
> ArgumentOutOfRangeException : buffersize is less than or equal to zero.

We can view the source code of the constructor to determine what the default buffer size would otherwise be.

File: src/GraphQLParser/GraphQLParser.csproj [hunk 5]
```diff
@@ -4,15 +4,17 @@
...
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="System.Memory" Version="4.5.5" Condition="'$(TargetFramework)' == 'netstandard2.0'" />
-    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.4" Condition="'$(TargetFramework)' == 'netstandard2.0'" />
+    <Using Remove="System.Net.Http" Condition="'$(TargetFramework)' == 'net462'" />
+    <PackageReference Include="System.ValueTuple" Version="4.5.0" Condition="'$(TargetFramework)' == 'net462'" />
+    <PackageReference Include="System.Memory" Version="4.5.5" Condition="'$(TargetFramework)' == 'netstandard2.0' or '$(TargetFramework)' == 'net462'" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.4" Condition="'$(TargetFramework)' == 'netstandard2.0' or '$(TargetFramework)' == 'net462'" />
```
Comment (Shane32) [inline src/GraphQLParser/GraphQLParser.csproj:17]: I would revert all changes here, so when we test against the .NET Framework, it uses the .NET Standard 2.0 TFM.  Note that if we do not revert these changes, we still do not have testing of the .NET Standard 2.0 build.  (We would then need to add testing against .NET Core 2.1 in order to test the .NET Standard 2.0 build.)

File: src/GraphQLParser/Visitors/SDLPrinterExtensions.cs (hunk 4 above)
Comment (Shane32) [reply src/GraphQLParser/Visitors/SDLPrinterExtensions.cs:38]: > add new method with bufferSize or change existing arguments in v10

I would not do either at this time.  Our API dictates that null can be passed to encoding, which will use UTF8 with no byte order mark.  Fine.  It doesn't say what buffer size we will use.  Fine.  No breaking changes.  No need to make changes.  Our code simply doesn't work as we documented that it should.

Comment (Shane32) [reply src/GraphQLParser/Visitors/SDLPrinterExtensions.cs:38]: Sorry, I stand corrected; we are following the .NET 6 specs; but that overload does not exist for .NET Standard 2.0.  So for .NET 6 it says the arguments are optional when in .NET Standard 2.0/2.1 they are required.  So we can pass `-1` and `null` for .NET 6, but for .NET Standard we must pass correct values.  So let's just use `#if !NET6_0` or whatever rather than adding a TFM to the build process for .NET framework.

Comment (Shane32) [reply src/GraphQLParser/Visitors/SDLPrinterExtensions.cs:38]: https://learn.microsoft.com/en-us/dotnet/api/system.io.streamreader.-ctor?view=net-6.0#system-io-streamreader-ctor(system-io-stream-system-text-encoding-system-boolean-system-int32-system-boolean)

https://learn.microsoft.com/en-us/dotnet/api/system.io.streamreader.-ctor?view=netstandard-2.1#system-io-streamreader-ctor(system-io-stream-system-text-encoding-system-boolean-system-int32-system-boolean)

Comment (sungam3r) [reply src/GraphQLParser/Visitors/SDLPrinterExtensions.cs]: Fine. Done.

File: src/GraphQLParser/GraphQLParser.csproj (hunk 5 above)
Comment (sungam3r) [reply src/GraphQLParser/GraphQLParser.csproj]: reverted

**Comment by sungam3r** [comment]
Hah.
```
Test run for /home/runner/work/parser/parser/src/GraphQLParser.Tests/bin/Debug/net462/GraphQLParser.Tests.dll (.NETFramework,Version=v4.6.2)
Microsoft (R) Test Execution Command Line Tool Version 17.7.1 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:02.58] GraphQLParser.Tests: Catastrophic failure: System.TypeLoadException: Could not load type of field 'Xunit.Runner.VisualStudio.VsExecutionSink:recorder' (4) due to: Could not load file or assembly 'Microsoft.VisualStudio.TestPlatform.ObjectModel, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
```

**Comment by sungam3r** [comment]
9.2.1 ?

**Comment by sungam3r** [comment]
I'm confused a bit. I just merged master into develop (actually fast-forwarded develop) and now master and develop both point to the same commit. What version to release - 8.4.1 or 9.2.1 ?

**Comment by sungam3r** [comment]
I'm done for today. You may go on and make release if you wish.

**Comment by Shane32** [comment]
Oh.  Sorry, we are using the `v8` branch for v8 and the `master` branch for v9

**Comment by Shane32** [comment]
I changed the default branch to `v8` so we don't make this mistake again.

**Comment by sungam3r** [comment]
Ok.
